Add rebuild button for PipelineRuns

a-roberts opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Adds a rebuild button for PipelineRuns. Takes an existing PipelineRun, does a lookup to find its actual object (from the CRD), modifies the name and omits the resource version (we'll get a new one on creation, definitely unique and handles for us), and kicks off a new PipelineRun based on it.

Added tests too that checks the label convention is met - the Go tests are the good ones really.

**The notification now appears on the PipelineRun page, right at the top. Rebuild button is now a ghost, off to the right, looking very swish**. 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

AlanGreene commented:
Rather than having a specific API for rebuilding a PipelineRun given a name, could we not have the client create the PipelineRun with the details extracted from the one being rebuilt?

a-roberts commented:
> Rather than having a specific API for rebuilding a PipelineRun given a name, could we not have the client create the PipelineRun with the details extracted from the one being rebuilt?

I wanted to that but it's easier here to handle the need for a new name and resource version, which i can handle trivially through this API (otherwise we'd do that through the frontend - and I'd rather put anything non-trivial manipulating resources on the backend)

jessm12 commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    /* eslint no-useless-rename: "error" */
+    const pipelinerunname = this.props.runName;
```
Consistent camel case please

jessm12 commented on pkg/endpoints/pipeline.go:
```diff
@@ -757,3 +763,138 @@ func (r Resource) UpdatePipelineRun(request *restful.Request, response *restful.
+	return newName
+}
+
+/* E.g. POST to http://localhost:9097/v1/namespaces/default/rebuild (provide name)
```
Provide name how?

jessm12 commented on pkg/endpoints/pipeline.go:
```diff
@@ -757,3 +763,138 @@ func (r Resource) UpdatePipelineRun(request *restful.Request, response *restful.
+			utils.RespondError(response, err, http.StatusInternalServerError)
+			return
+		} else {
+			// ALl is well, include the name of the new rebuilt run in the response
```
typo

jessm12 commented on pkg/endpoints/pipeline.go:
```diff
@@ -757,3 +763,138 @@ func (r Resource) UpdatePipelineRun(request *restful.Request, response *restful.
+		}
+	}
+	// We should never get here - but if so, they've given us a request we can't handle
+	logging.Log.Error("didn't receive a request we could actually handle!")
```
Maybe log the status message too

jessm12 commented on pkg/endpoints/pipeline_test.go:
```diff
@@ -1167,3 +1167,137 @@ func TestPOSTCreatePipelineRuns(t *testing.T) {
+	pipeline := createPipelineForTesting(testPipelineNameToRun, namespace, server, r, t)
+	pipelineRun := createPipelineRunForTesting(testPipelineNameToRun, namespace, server, r, t)
+
+	// Check we've got a new PipelineRun and its the same apart from name and resource version
```
this comment confused me, isn't this supposed to be on the second creation only?

jessm12 commented:
![Screenshot 2019-07-04 at 17 21 53](https://user-images.githubusercontent.com/32932155/60679640-3ee97300-9e80-11e9-9c43-bea07dbcbbb4.png)

this doesn't look quite right 🤔

a-roberts commented:
^^ I'll fix that, trying low contrast 😄  and moving to the right, then margins

a-roberts commented:
Working on displaying the notification on the PipelineRun page itself now and then I'm done with this

a-roberts commented:
@AlanGreene @ncskier could I get another review please?

![image](https://user-images.githubusercontent.com/12814972/60985603-a8acc580-a335-11e9-958a-ed02193ad653.png)

And we get a toast notification now above, on that same page. I think my Go tests are decent but the React ones less so - but we're thinking we may change the PipelineRun code big time soon for the dash anyway. Open to improvement suggestions, not sure if I'm correct in passing the functions as props from the parent to child so I can set state accordingly.

👼

AlanGreene commented on Rebuild.scss:
```diff
@@ -0,0 +1,5 @@
+#rebuild-btn {
```
this file should be removed

AlanGreene commented on src/components/Rebuild/Rebuild.test.js:
```diff
@@ -0,0 +1,53 @@
+    .spyOn(API, 'rebuildPipelineRun')
+    .mockImplementation(() => Promise.resolve(headers));
+  renderWithRouter(<Rebuild {...props} runName="fake-pipeline-run" />);
+  const rebuildButton = document.getElementById('rebuild-btn');
```
prefer `getByTestId` or similar functions from react-testing-library#render over using document.getElementById and other native DOM functions

AlanGreene commented on src/components/Rebuild/Rebuild.test.js:
```diff
@@ -0,0 +1,53 @@
+  }
+};
+
+it('rebuild button creates API call with correct run name', async () => {
```
doesn't look like `async` is needed here

AlanGreene commented on src/components/Rebuild/Rebuild.scss:
```diff
@@ -0,0 +1,11 @@
+.rebuild-btn {
+  position: absolute !important;
```
avoid `!important`, instead prefer more specific selectors

`.bx--btn.rebuild-btn` or similar


also avoid `position: absolute` if possible as it can cause problems with responsive layouts, leading to overlapping content and other issues

AlanGreene commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /* eslint no-useless-rename: "error" */
+    const pipelinerunname = this.props.runName;
```
why are these eslint comments needed?

a-roberts replied:
Destructuring occurs otherwise so if we were to run `npm run lint:fix`, the variable name would become runName instead. I really do want it to be `pipelinerunname` as this is to be used in the JSON payload for the actual rebuild function

I'll add this as a comment in the code to make it obvious for maintainers too

AlanGreene replied:
Rather than renaming it here you should be able to do it where it's used in the payload below and avoid these eslint comments. Normally the comments are used to allow something that would otherwise be a violation, it's odd to see it used this way.

AlanGreene replied:
I'd like to revisit these, but not a blocker for this PR.

AlanGreene commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,86 @@
+    const pipelinerunname = this.props.runName;
+    /* eslint no-useless-rename: "error" */
+    const pipelineName = this.props.pipelineRun.spec.pipelineRef.name;
+    const { namespace } = this.props.pipelineRun.metadata.namespace; // was namespaceToUse and OK
```
this doesn't look right, `namespace.namespace`...

a-roberts replied:
Yeah, fixed now, and my test code that was expecting it to be called with undefined and not the actual namespace, thanks

AlanGreene commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,86 @@
+
+    const promise = rebuildPipelineRun(namespace, payload);
+
+    promise
```
no need for the `const` here, may have been leftover from previous code / debugging

```js
rebuildPipelineRun(namespace, payload)
  .then(header => {
```

AlanGreene commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,86 @@
+          logsURL.lastIndexOf('/') + 1
+        );
+        const finalURL = `/namespaces/${namespace}/pipelines/${pipelineName}/runs/${newPipelineRunName}`;
+        this.props.setShowRebuildNotification(true);
```
Rather than setting these individually can we just set a single object? Something like this:

```js
this.props.showRebuildNotification({
  message: 'Rebuilt run successfully',
  kind: 'success',
  logsURL
});
```

In PipelineRun you would `setState({ rebuildNotification })` with this object as the value.

a-roberts replied:
Great idea, trying

a-roberts replied:
Done, this was a nice refactor, thank you

AlanGreene commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,86 @@
+
+  render() {
+    return (
+      <>
```
no need for a fragment here as it's just returning a single button

AlanGreene commented on src/components/Rebuild/Rebuild.js:
```diff
@@ -0,0 +1,86 @@
+      <>
+        <Button
+          data-testid="rebuild-btn"
+          id="rebuild-btn"
```
see other comment regarding use of testid, `id` shouldn't be necessary

AlanGreene commented on src/components/RunHeader/RunHeader.js:
```diff
@@ -44,16 +44,23 @@ class RunHeader extends Component {
-                <span className="status-label">{reason}</span>
-                <span className="time">{lastTransitionTime}</span>
-              </h1>
+              <div>
```
Is the wrapping `div` here used to ensure the block extends the full width? If so it would be better to achieve this via explicit styles in the SCSS rather than adding the extra wrapper, this will make it easier to handle the responsive layout at smaller screen sizes.

a-roberts replied:
I'll run with this as well

a-roberts replied:
I think I added this so that it was all in one div to return as a jsx element and it wasn't to do with the block - but I don't even need it

Just built and tested, looks great without it anyway so I've removed it

AlanGreene commented on src/components/RunHeader/RunHeader.js:
```diff
@@ -44,16 +44,23 @@ class RunHeader extends Component {
+                  <span className="status-label">{reason}</span>
+                  <span className="time">{lastTransitionTime}</span>
+                  <Rebuild
+                    {...this.props}
```
Does the Rebuild button require all of the props from RunHeader or just a subset?

a-roberts replied:
I'm passing 

```
    setShowRebuildNotification={this.setShowRebuildNotification}
          setShowRebuildNotificationKind={this.setShowRebuildNotificationKind}
          setShowRebuildNotificationLogsURL={this.setShowRebuildNotificationLogsURL}
          setShowRebuildNotificationMessage={this.setShowRebuildNotificationMessage}
```

to the RunHeader from PipelineRun which Rebuild needs to access, so unfortunately yeah 😢

AlanGreene commented on src/containers/PipelineRun/PipelineRun.js:
```diff
@@ -225,13 +272,52 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
+          />
+        )}
+
+        <div>
```
what's the purpose of this additional `div` wrapper?

AlanGreene commented on src/containers/PipelineRun/PipelineRun.js:
```diff
@@ -225,13 +272,52 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
-          reason={pipelineRunReason}
-          status={pipelineRunStatus}
-        />
+        {/* prettier-ignore */ (showRebuildNotification && !showRebuildNotificationLogsURL) && (
```
Prettier complaining about unnecessary parentheses? Let's just keep it happy 😉

a-roberts replied:
Prettier removes the brackets which I think changes the functionality as it doesn't evaluate both expressions, lemme check to confirm...

a-roberts replied:
My bad, removed the rule, all is well

AlanGreene commented on src/containers/PipelineRun/PipelineRun.js:
```diff
@@ -225,13 +272,52 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
+            caption={
+              <Link id="newpipelinerunlink" 
+                to={showRebuildNotificationLogsURL}
+                onClick={ () => this.setShowRebuildNotification(false)}
```
if you're navigating away and this component is unmounted the state will be re-initialised next time you return anyway

a-roberts replied:
Hmm without this added, on the rebuilt PipelineRun this caption is presented

AlanGreene replied:
my bad, since you're navigating from PipelineRun to another PipelineRun the component isn't unmounted just updated

a-roberts commented:
@AlanGreene ready for another review please, I think I've addressed all of your comments (e.g. the props refactor and selective CSS without using !important). I fixed some impl code and thus test code along the way too for the namespace reference.

AlanGreene commented on src/components/Rebuild/Rebuild.scss:
```diff
@@ -0,0 +1,6 @@
+//.rebuild-btn {
```
remove

AlanGreene commented on src/components/RunHeader/RunHeader.js:
```diff
@@ -53,6 +53,11 @@ class RunHeader extends Component {
                 </div>
                 <span className="status-label">{reason}</span>
                 <span className="time">{lastTransitionTime}</span>
+                <Rebuild
```
RunHeader is reused for TaskRuns, and although it would be useful to have there it doesn't work at the moment so I'd suggest adding this only when `pipelineRun` is set.

a-roberts replied:
On it, I'll add
```
      {pipelineRun && (
                  <Rebuild
                    {...this.props}
                    pipelineRun={pipelineRun}
                    runName={runName}
                  />
                )}
```
and retest

a-roberts commented:
@AlanGreene ready for another one please 😄
